rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp

```diff
@@ -402,30 +420,61 @@ void SqliteStorage::prepare_for_reading()
         topic_list += ",";
       }
     }
-    statement_str += "(topics.name IN (" + topic_list + ")) AND ";
+    where_conditions.push_back("(topics.name IN (" + topic_list + "))");
   }
   // add topic filter based on regular expression
   if (!storage_filter_.topics_regex.empty()) {
     // Construct string for selected topics
-    statement_str += "(topics.name REGEXP '" + storage_filter_.topics_regex + "')";
-    statement_str += " AND ";
+    where_conditions.push_back("(topics.name REGEXP '" + storage_filter_.topics_regex + "')");
   }
   // exclude topics based on regular expressions
   if (!storage_filter_.topics_regex_to_exclude.empty()) {
     // Construct string for selected topics
-    statement_str += " (topics.name NOT IN ";
-    statement_str += " (SELECT topics.name FROM topics WHERE topics.name REGEXP '";
-    statement_str += storage_filter_.topics_regex_to_exclude + "')";
-    statement_str += " ) AND ";
+    where_conditions.push_back(
+      "(topics.name NOT IN "
+      "(SELECT topics.name FROM topics WHERE topics.name REGEXP '" +
+      storage_filter_.topics_regex_to_exclude + "'))");
+  }
+
+  const std::string direction_op = read_order_.reverse ? "<" : ">";
+  const std::string order_direction = read_order_.reverse ? "DESC" : "ASC";
+
+  // add seek head filter
+  if (read_order_.sort_by == rosbag2_storage::ReadOrder::File) {
+    where_conditions.push_back(
+      "(messages.id " + direction_op + "= " + std::to_string(seek_row_id_) + ") ");
+  } else {
+    where_conditions.push_back(
+      "(((timestamp = " + std::to_string(seek_time_) + ") "
+      "AND (messages.id " + direction_op + "= " + std::to_string(seek_row_id_) + ")) "
+      "OR (timestamp " + direction_op + " " + std::to_string(seek_time_) + ")) ");
```

Comment:
We have enum values for read order ether by `File` or by `Timestamp`, having sort of mixed read order by file and by timestamp will be unexpected. For consistency I would replace those `OR` `AND` with strict ordering by timestamp even though we had such weird sorting in baseline.
At least it will be consistent and expected behavior.
```suggestion
      "(timestamp " + direction_op + "= " + std::to_string(seek_time_) + ") ");
```

Reply:
(correct me if my analysis here is wrong, but) This baseline ordering exists for a reason and is not weird or mixed. It enforces id-ordering between messages with the same timestamp. Timestamp is not a unique field, but `rowid` is. This would be fine if we never reset the query, because we could just `++current_message_row_` iterator from the query, but many operations can reset the query without changing the seek head, meaning we have to track both ordering fields to reach the correct place in the iteration.

Take for example the following data:
```
id: timestamp
1: 100
2: 100
3: 200
```

Now let's walk through the following scenario:
* `set_order(TIMESTAMP)`
* `read_next`
* `set_filter` (or any operation that doesn't change the current place but resets the query)
* `read_next`
* `set_filter`
* `read_next`

With current logic:
* `seek_row_id_ = 0, seek_time_ = 0`
* `read_next` -> `(timestamp = 0 and id >= 0)` -> returns `1: 100`.
* update: `seek_row_id_ = 1; seek_time_ = 100`
* new query created
* `read_next` -> (timestamp = 100 and id >= 1)` -> return `2: 100`
* update `seek_row_id_ = 1; 
* new query created
* `read_next` -> `(timestamp = 100 and id >= 2)` (no data) `OR timestamp > 100` -> return `3: 200`

With your proposal:
* `seek_row_id_ = 0, seek_time_ = 0`
* `read_next` -> `timestamp >= 0` -> returns `1: 100`
* update: `seek_time_ = 100`
* new query created
* `read_next` -> `timestamp >= 100` -> returns `1: 100`
* etc... returns the same message forever

Feel free to test it out yourself, but I'm pretty sure it just doesn't work.

Reply:
@emersonknapp Thanks for explanation and detailed examples :+1: 
Now I do see that row_id is matter. 
And it's matter because we allow assumption that index could be not unique in this case we sort of indexing  by "timestamp". 
From real word scenario in very rare cases it could be that timestamp not unique. And this assumption looks reasonable.

Could you please add comment somewhere that we do this mixed ordering because have assumption that timestamp not unique and could be two records with the same timestamp and we will stuck with `read_next` in this case?

Reply:
I have added this comment